pkg/operations/kubernetesupgrade/upgradeagentnode.go

```diff
@@ -70,7 +70,11 @@ func (kan *UpgradeAgentNode) DeleteNode(vmName *string, drain bool) error {
 	}
 	// Cordon and drain the node
 	if drain {
-		err = operations.SafelyDrainNodeWithClient(client, kan.logger, nodeName, kan.cordonDrainTimeout)
+		if kan.UpgradeContainerService.Properties.IsAzureStackCloud() {
```

Comment:
Will this work?

```
err = operations.SafelyDrainNodeWithClient(client, kan.logger, nodeName, kan.cordonDrainTimeout)
if kan.UpgradeContainerService.Properties.IsAzureStackCloud() {
    err = operations.WaitForDisksAttached(client, kan.logger, nodeName, kan.cordonDrainTimeout)
}
```

I think it is slightly better to compose discrete operations and unit testing will likely be easier.

Reply:
Yes, actually this was my first attempt. I totally agree that it's easier to unit-test. However, let me explain the reason why I changed my mind. So inside the function SafelyDrainNodeWithClient things happen in this order: 1) mark the node as unschedulable, 2) get the list of pods that need to be evicted from this node, 3) evict these pods. And for the volume check, one more step is added: 4) check the volumes for the evicted pods. So the point is, in order to check/monitor the volumes, I need to get a list of pods that are evicted from this node. It's not easy to properly get these pods from outside of the function SafelyDrainNodeWithClient.
The alternatives would be 1) get the pods before and after node drain and do the diff to get the pod that needs to be monitored, or 2) change the signature of function SafelyDrainNodeWithClient to make it return the evicted pods. I tried both of these two approaches which worked as well. Please advise if you have a better solution.